tests/integration/test_forward_auth.py

```diff
@@ -0,0 +1,187 @@
+        trust=True,
+    )
+
+    await ops_test.model.integrate(f"{IAP_REQUIRER_CHARM}:ingress", TRAEFIK_CHARM)
+    await ops_test.model.integrate(f"{IAP_REQUIRER_CHARM}:auth-proxy", OATHKEEPER_CHARM)
+
+    await ops_test.model.integrate(f"{TRAEFIK_CHARM}:experimental-forward-auth", OATHKEEPER_CHARM)
+
+    await ops_test.model.wait_for_idle(
+        [TRAEFIK_CHARM, OATHKEEPER_CHARM, IAP_REQUIRER_CHARM], status="active", timeout=1000
+    )
+
+
+@retry(
+    wait=wait_exponential(multiplier=3, min=1, max=10),
+    stop=stop_after_attempt(5),
+    reraise=True,
+)
+async def test_allowed_forward_auth_url_redirect(ops_test: OpsTest) -> None:
+    """Test that a request hitting an application protected by IAP is forwarded by traefik to oathkeeper.
+
+    An allowed request should be performed without authentication.
+    """
+    requirer_url = await get_reverse_proxy_app_url(ops_test, TRAEFIK_CHARM, IAP_REQUIRER_CHARM)
+
+    protected_url = join(requirer_url, "anything/allowed")
+
+    resp = requests.get(protected_url, verify=False)
+    assert resp.status_code == 200
+
+
+async def test_protected_forward_auth_url_redirect(ops_test: OpsTest) -> None:
+    """Test that when trying to reach a protected url, the request is forwarded by traefik to oathkeeper.
+
+    An unauthenticated request should then be denied with 401 Unauthorized response.
+    """
+    requirer_url = await get_reverse_proxy_app_url(ops_test, TRAEFIK_CHARM, IAP_REQUIRER_CHARM)
+
+    protected_url = join(requirer_url, "anything/deny")
+
+    resp = requests.get(protected_url, verify=False)
+    assert resp.status_code == 401
+
+
+async def test_forward_auth_url_response_headers(
+    ops_test: OpsTest, lightkube_client: Client
+) -> None:
+    """Test that a response mutated by oathkeeper contains expected custom headers."""
+    requirer_url = await get_reverse_proxy_app_url(ops_test, TRAEFIK_CHARM, IAP_REQUIRER_CHARM)
+    protected_url = join(requirer_url, "anything/anonymous")
+
+    # Push an anonymous access rule as a workaround to avoid deploying identity-platform bundle
+    anonymous_rule = [
+        {
+            "id": "iap-requirer:anonymous",
+            "match": {
+                "url": protected_url,
+                "methods": ["GET", "POST", "OPTION", "PUT", "PATCH", "DELETE"],
+            },
+            "authenticators": [{"handler": "anonymous"}],
+            "mutators": [{"handler": "header"}],
+            "authorizer": {"handler": "allow"},
+            "errors": [{"handler": "json"}],
+        }
+    ]
+
+    update_access_rules_configmap(ops_test, lightkube_client, rule=anonymous_rule)
+    update_config_configmap(ops_test, lightkube_client)
+
+    assert_anonymous_response(protected_url)
+
+
+@retry(
+    wait=wait_exponential(multiplier=3, min=1, max=10),
+    stop=stop_after_attempt(10),
+    reraise=True,
+)
+def assert_anonymous_response(url):
+    resp = requests.get(url, verify=False)
+    assert resp.status_code == 200
```

Comment:
if we used an async client (aiohttp) we could drop the retry bit in this case
we'd still be blocking here though https://github.com/canonical/traefik-k8s-operator/pull/275/commits/13f93cae41764bab15e7e5ad48f228b5482e8322#diff-03198ed59d11f58b5b2b25d32d96a8726c03c0fe4b25df59e1cc7cbefad31e8eR160